### PR TITLE
#DP-64 ft: OrgsignUp and login

### PR DESCRIPTION
**Pull request by jeandamascyifuzo**
### PR Description
Organisation and User signup

### Description of tasks that were expected to be completed
- [x] An admin should be able to login
- [ ] An admin should be able to add/register a new org
- [x] Users are able to login
- [x] Users should be presented with their respective interfaces

### Functionality
```
git clone https://github.com/atlp-rwanda/atlp-pulse-bn.git
cd atlp-pulse-bn
git checkout ft-OrgSignUp-dp-64
docker run --name devpulse_db_dev -p 27000:27017 -d mongo:latest
npm i 
npm run dev
```
### Testing mutation
```
e.g: mutation Mutation($registerInput: RegisterInput) {
  createUser(registerInput: $registerInput) {
    user {
      email
      id
      role
    }
  }
}

field: {
  "registerInput": {
    "email": "",
    "password": "",
    "role": ""
  }
}

mutation LoginUser($loginInput: LoginInput) {
  loginUser(loginInput: $loginInput) {
    token
    user {
      email
      role
    }
  }
}
filed:
{
  "loginInput": {
    "email": "",
    "password": "",
  }
}
```

### CODE REVIEW

File: src/resolvers/resolver.ts [hunk 1]
```diff
@@ -1,26 +1,64 @@
...
+      });
+      const token = jwt.sign(
+        { userId: newUser._id, role: newUser?.role },
+        SECRET,
+        { expiresIn: "2h" }
+      );
+
+      return { token, user: newUser };
+    },
+    async loginUser(_: any, { loginInput: { email, password } }: any) {
```
Comment (chanceDira) [inline src/resolvers/resolver.ts]: After creating a user and assigning him/her a role, like an admin role. the user is always authenticated as a "user" on login

File: src/resolvers/resolver.ts [hunk 2]
```diff
@@ -1,26 +1,64 @@
...
+      const user = await User.findOne<any>({ where: { email: email } });
+
+      if (await user?.checkPass(password)) {
```
Comment (chanceDira) [inline src/resolvers/resolver.ts]: we can also check the email before generating the token. I provided a wrong email and the right password and I was able to access the token. You can verify if the token is generated upon valid email and password.

**Review by chanceDira — CHANGES_REQUESTED** [review]
Nice work @jeandamascyifuzo,

Below you can find some feedback.

File: src/resolvers/resolver.ts (hunk 2 above)
Comment (feyton) [reply src/resolvers/resolver.ts:41]: Chance how can we reproduce the scenario of this issue.

Comment (feyton) [reply src/resolvers/resolver.ts]: @chance

how can we reproduce the scenario of this issue.

**Review by chanceDira — APPROVED** [review]
Nice Job @jeandamascyifuzo @chanceDira

**Review by feyton — APPROVED** [review]
@jeandamascyifuzo 

This really a great work

**Review by rukundo-kevin — APPROVED** [review]

**Review by PatrickTUy — APPROVED** [review]
Approved

File: src/models/user.ts [hunk 3]
```diff
@@ -1,49 +1,55 @@
...
+        firstName: {
+            type: String,
+            required: true,
+        },
+        lastName: {
+            type: String,
+            required: true,
+        },
+        address: {
+            type: String,
```
Comment (ceelogre) [inline src/models/user.ts:44]: I see changes related to indentation. There's no team-wide linting config?

File: src/resolvers/resolver.ts [hunk 4]
```diff
@@ -1,31 +1,67 @@
...
 const resolvers = {
   Query: {
-    hello: () => "Hellooo, welcome to your Graphql server",
-    async getAllUsers() {
-      const users = await User.find({});
-      return users;
-    },
-    async getAllProfiles() {
+    async getAllProfiles(_: any, _args: any, context: { userId: any }) {
+      if (!context.userId) throw new Error("Unauthorized");
```
Comment (ceelogre) [inline src/resolvers/resolver.ts]: What part of the code catches this error when thrown?

**Review by ceelogre — CHANGES_REQUESTED** [review]
Emails are not validated.
image

Roles should be one of the enum values. [admin, trainee,...]
image



Also, let's not return the error stack trace to the user. Be wary of throwing errors without catch blocks. Look at my comments.

**Comment by chanceDira** [comment]
> Emails are not validated. image
> 
> Roles should be one of the enum values. [admin, trainee,...] image
> 
> Also, let's not return the error stack trace to the user. Be wary of throwing errors without catch blocks. Look at my comments.

@ceelogre I have the task of managing roles, what I do to assign roles is first to create roles apart and then assign a role that exists in the role's table. so my Idea here. @jeandamascyifuzo can first check if the role is available in the role's table if yes use it if not throws an error.

File: src/models/user.ts (hunk 3 above)
Comment (jeandamascyifuzo) [reply src/models/user.ts:44]: well resolved

File: src/resolvers/resolver.ts (hunk 4 above)
Comment (jeandamascyifuzo) [reply src/resolvers/resolver.ts]: am focusing on authentication not profile
may be such issue will be resolved later

**Comment by jeandamascyifuzo** [comment]
@ceelogre 
what will happen if the organisation doesn't use such roles [admin, coordinators, trainees],
My idea is that let's create role apart and then assign that role to the user instead of hardcoding them.
@chanceDira is doing that task of managing the role >

**Review by ceelogre — APPROVED** [review]

**Pull request merged**